### PR TITLE
Add Dapr integration to Tye

### PR DESCRIPTION
**Pull request by rynowak**
Adds a new `extensions` integration point. Currently all extensbility has
to be inside the Tye codebase. There's no functionality for loading or distributing
plugins.

Enable dapr for an application like:

```yaml
name: test_app
extensions:
- name: dapr
```

The dapr extension currently accepts and requires no additional options, and applies
to all services defined in the application.

What it does:

- In local run: starts a dapr sidecar for each project that does http
- In local run: sidecars start in the directory of tye.yaml, so dapr will look for
  component manifests in ./components (relative to tye.yaml
- In deployment: adds required annotations to deployments

Some new features that were added to tye to enable this:

- Config: Ability to parse and conditionally run extensions
- Run: Ability to token replace env-vars into command line args
- Deploy: Ability to configure labels and annotations

### CODE REVIEW

File: src/Microsoft.Tye.Extensions/Microsoft.Tye.Extensions.csproj [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (rynowak) [inline src/Microsoft.Tye.Extensions/Microsoft.Tye.Extensions.csproj]: New project to keep us from leaking "extensions" code into the core.

File: samples/dapr/.tye/process_store [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{"pid":"80819"}
```
Comment (davidfowl) [inline samples/dapr/.tye/process_store]: Remove? We should add .tye to .gitignore

File: src/Microsoft.Tye.Extensions/Dapr/DaprExtension.cs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+                    var metrics = new BindingBuilder()
+                    {
+                        AutoAssignPort = true,
+                        Name = "metrics",
+                        Protocol = "http",
+                    };
+                    proxy.Bindings.Add(metrics);
+
+                    // Set APP_PORT based on the project's assigned port for http
+                    if (httpBinding == null)
```
Comment (davidfowl) [inline src/Microsoft.Tye.Extensions/Dapr/DaprExtension.cs]: Didn't we already check this for null?

File: samples/dapr/.tye/process_store (hunk 2 above)
Comment (rynowak) [reply samples/dapr/.tye/process_store]: yes. We should consider moving this to user profile if we have to as well.

**Comment by davidfowl** [comment]
Small nits but LGTM

File: samples/dapr/.tye/process_store (hunk 2 above)
Comment (davidfowl) [reply samples/dapr/.tye/process_store]: Sure the only thing I’m worried about is that it needs to be keyed by something so we can find it on the next run. If we don’t multiple runs by the same user can conflict

Comment (rynowak) [reply samples/dapr/.tye/process_store]: A pretty simple way to do that is to key based on PID of the Host. So you get a hierarchy like:

```
$Home/.tye/pid/12345/<data we capture today>
$Home/.tye/pid/56789/<data we capture today>
```

Then the host can look for orphaned processed from another run by considering:

- pid of other running host process
- folders that exist
- (optional) timestamps of files

The only case where you'd have a problem is if the OS decided to reuse a PID for another tye host process.

**Pull request merged**